src/Course2/week1/TopologicalSorting.java

```diff
@@ -0,0 +1,61 @@
+package Course2.week1;
+import java.util.*;
+import java.util.concurrent.CopyOnWriteArrayList;
+
+public class TopologicalSorting {
+
+    static void findTopoSort(int node, int vis[], ArrayList<ArrayList<Integer>> adj, Stack<Integer> st) {
+        vis[node] = 1;
+        for(Integer it: adj.get(node)) {
+            if(vis[it] == 0) {
+                findTopoSort(it, vis, adj, st);
+            }
+        }
+        st.push(node);
+    }
+    static int[] topoSort(int N, ArrayList<ArrayList<Integer>> adj) {
+        Stack<Integer> st = new Stack<Integer>();
+        int vis[] = new int[N];
+
+        for(int i = 0;i<N;i++) {
+            if(vis[i] == 0) {
+                findTopoSort(i, vis, adj, st);
+            }
+        }
+
+        int topo[] = new int[N];
+        int ind = 0;
+        while(!st.isEmpty()) {
+            topo[ind++] = st.pop();
+        }
+
+        return topo;
+    }
+    public static void main(String args[])
+    {
+        ArrayList<ArrayList<Integer>> adj=new ArrayList<>();
+        int n=6;
+        for(int i=0;i<n;i++)
+        {
+            ArrayList<Integer> arr=new ArrayList<>();
```

Comment:
limit arraylist size to 1/2, if you only need to add 1/2 elements or use array instead